hack/generator/pkg/astmodel/resource.go

```diff
@@ -0,0 +1,131 @@
+/*
+ * Copyright (c) Microsoft Corporation.
+ * Licensed under the MIT license.
+ */
+
+package astmodel
+
+import (
+	"go/ast"
+	"go/token"
+)
+
+// CreateResourceDefinitions creates definitions for a resource
+func CreateResourceDefinitions(name *TypeName, specType *StructType, statusType *StructType, idFactory IdentifierFactory) (TypeDefiner, []TypeDefiner) {
+
+	var others []TypeDefiner
+
+	var specName *TypeName
+	if specType != nil {
+		specName = NewTypeName(name.PackageReference, name.Name()+"Spec")
+		specDef, specOthers := specType.CreateDefinitions(specName, idFactory)
+
+		others = append(append(others, specDef), specOthers...)
+	} else {
+		panic("spec must always be provided")
+	}
+
+	var statusName *TypeName
+	if statusType != nil {
+		statusName = NewTypeName(name.PackageReference, name.Name()+"Status")
+		statusDef, statusOthers := statusType.CreateDefinitions(statusName, idFactory)
+
+		others = append(append(others, statusDef), statusOthers...)
+	}
+
+	this := &ResourceDefinition{typeName: name, spec: specName, status: statusName, isStorageVersion: false}
+
+	return this, others
+}
+
+// ResourceDefinition represents an ARM resource
+type ResourceDefinition struct {
+	typeName         *TypeName
+	spec             *TypeName
+	status           *TypeName
+	isStorageVersion bool
+	description      *string
+}
+
+// assert that ResourceDefinition implements TypeDefiner
+var _ TypeDefiner = &ResourceDefinition{}
+
+func (definition *ResourceDefinition) Name() *TypeName {
+	return definition.typeName
+}
+
+func (definition *ResourceDefinition) Type() Type {
+	return definition.spec // TODO?????
+}
+
+func (definition *ResourceDefinition) WithIsStorageVersion(isStorageVersion bool) *ResourceDefinition {
+	result := *definition
+	result.isStorageVersion = isStorageVersion
+	return &result
+}
```

Comment:
The `WithXYZ()` syntax is working well, but given this is a boolean parameter, elsewhere I've used `MarkAsXYZ` instead.

```suggestion
func (definition *ResourceDefinition) MarkAsStorageVersion() *ResourceDefinition {
	result := *definition
	result.isStorageVersion = true
	return &result
}
```

We might not even need `MarkAsNonStorageVersion()`.